applications/HDF5Application/custom_io/hdf5_file.cpp

```diff
@@ -662,272 +622,229 @@ void File::SetEchoLevel(int Level)
-    KRATOS_ERROR_IF(attr_type_id < 0) << "H5Aget_type failed." << std::endl;
-    htri_t is_valid_type = H5Tequal(mem_type_id, attr_type_id);
-    KRATOS_ERROR_IF(H5Tclose(attr_type_id) < 0) << "H5Tclose failed." << std::endl;
-    KRATOS_ERROR_IF(is_valid_type < 0) << "H5Tequal failed." << std::endl;
-    KRATOS_ERROR_IF(is_valid_type == 0) << "Memory and file data types are different." << std::endl;
+    type_traits::Reshape(rValue, shape);
 
-    // Check dimensions.
-    space_id = H5Aget_space(attr_id);
-    KRATOS_ERROR_IF(space_id < 0) << "H5Aget_space failed." << std::endl;
-    KRATOS_ERROR_IF((ndims = H5Sget_simple_extent_ndims(space_id)) < 0)
-        << "H5Sget_simple_extent_ndims failed." << std::endl;
-    KRATOS_ERROR_IF(ndims != 2) << "Attribute \"" << rName << "\" is not matrix." << std::endl;
-    KRATOS_ERROR_IF(H5Sget_simple_extent_dims(space_id, dims, nullptr) < 0)
-        << "H5Sget_simple_extent_dims failed" << std::endl;
-    KRATOS_ERROR_IF(H5Sclose(space_id) < 0) << "H5Sclose failed." << std::endl;
-    rValue.resize(dims[0], dims[1], false);
     // Read attribute.
-    KRATOS_ERROR_IF(H5Aread(attr_id, mem_type_id, &rValue(0,0)) < 0) << "H5Aread failed." << std::endl;
-    KRATOS_ERROR_IF(H5Aclose(attr_id) < 0) << "H5Aclose failed." << std::endl;
+    KRATOS_ERROR_IF(H5Aread(attr_id, mem_type_id, type_traits::GetContiguousData(rValue)) < 0)
+        << "H5Aread failed." << std::endl;
+    KRATOS_ERROR_IF(H5Aclose(attr_id) < 0)
+        << "H5Aclose failed." << std::endl;
+
     KRATOS_INFO_IF("HDF5Application", GetEchoLevel() == 2)
         << "Read time \"" << rObjectPath << '/' << rName
         << "\": " << timer.ElapsedSeconds() << std::endl;
+
     KRATOS_CATCH("Path: \"" + rObjectPath + '/' + rName + "\".");
 }
 
-void File::ReadAttribute(const std::string& rObjectPath, const std::string& rName, std::string& rValue)
+Parameters File::ReadAttribute(const std::string& rObjectPath)
 {
-    KRATOS_TRY;
-    BuiltinTimer timer;
-    hid_t mem_type_id, attr_type_id, space_id, attr_id;
-    int ndims;
-    hsize_t dims[2];
-    const unsigned max_ssize = 100;
-    char buffer[max_ssize];
+    KRATOS_TRY
 
-    mem_type_id = H5T_NATIVE_CHAR;
-    attr_id = H5Aopen_by_name(m_file_id, rObjectPath.c_str(), rName.c_str(),
-                                    H5P_DEFAULT, H5P_DEFAULT);
-    KRATOS_ERROR_IF(attr_id < 0) << "H5Aopen_by_name failed." << std::endl;
+    Parameters result;
 
-    // Check data type.
-    attr_type_id = H5Aget_type(attr_id);
-    KRATOS_ERROR_IF(attr_type_id < 0) << "H5Aget_type failed." << std::endl;
-    htri_t is_valid_type = H5Tequal(mem_type_id, attr_type_id);
-    KRATOS_ERROR_IF(H5Tclose(attr_type_id) < 0) << "H5Tclose failed." << std::endl;
-    KRATOS_ERROR_IF(is_valid_type < 0) << "H5Tequal failed." << std::endl;
-    KRATOS_ERROR_IF(is_valid_type == 0) << "Attribute \"" << rName << "\" is not a string." << std::endl;
+    const auto& attribute_names = GetAttributeNames(rObjectPath);
```

Comment:
```suggestion
    const auto attribute_names = GetAttributeNames(rObjectPath);
```

I know it's the same thing, but it's confusing and it rapes my brain whenever I see a temporary captured as a ref.

Reply:
I would keep it, because then we don't have to know what kind of a return is given by the method. If it is temporary, it is the same, if it is not a temporary, then it gets the const ref from the method.